File: ios/engine/KMEI/KeymanEngine/Classes/LanguagePicker/LanguageDetailViewController.m
//
//  LanguageDetailViewController.m
//  Keyman Engine
//
//  Created by Serkan Kurt on 26/07/13.
//  Copyright (c) 2017 SIL International. All rights reserved.
//

#import "LanguageDetailViewController.h"
#import "KeyboardPickerViewController.h"
#import "KeyboardNameTableViewCell.h"
#import "KMManager+Internal.h"

static const NSInteger kmToolbarButtonTag = 100;
static const NSInteger kmToolbarLabelTag = 101;
static const NSInteger kmToolbarActivityIndicatorTag = 102;

@interface LanguageDetailViewController ()
    @property (nonatomic, strong) NSMutableDictionary *userKeyboards;
    @property (assign) BOOL isUpdate;
@end

@implementation LanguageDetailViewController

@synthesize userKeyboards, languageIndex, languageName, languageID, keyboards;

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)loadView {
	[super loadView];
    [self loadUserKeyboards];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardDownloadStarted:) name:kKeymanKeyboardDownloadStartedNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardDownloadFailed:) name:kKeymanKeyboardDownloadFailedNotification object:nil];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
}

- (void)dealloc {
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [keyboards count];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *cellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (!cell) {
        cell = [[KeyboardNameTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:cellIdentifier];
        UIView *selectionColor = [[UIView alloc] init];
        [selectionColor setBackgroundColor:[UIColor colorWithRed:204.0/255.0 green:136.0/255.0 blue:34.0/255.0 alpha:1.0]];
        [cell setSelectedBackgroundView:selectionColor];
    }
    
    return cell;
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
	//[cell setSelectionStyle:UITableViewCellSelectionStyleNone];
    [(KeyboardNameTableViewCell *)cell setIndexPath:indexPath];
    
    cell.textLabel.text = [[keyboards objectAtIndex:indexPath.section] objectForKey:kKeymanNameKey];
    NSString *keyboardID = [[keyboards objectAtIndex:indexPath.section] objectForKey:kKeymanIdKey];
    if ([self isAdded:languageID keyboardID:keyboardID]) {
        [cell setAccessoryType:UITableViewCellAccessoryCheckmark];
        cell.userInteractionEnabled = NO;
        cell.textLabel.enabled = NO;
        cell.detailTextLabel.enabled = NO;
    }
    else {
        [cell setAccessoryType:UITableViewCellAccessoryNone];
        cell.userInteractionEnabled = YES;
        cell.textLabel.enabled = YES;
        cell.detailTextLabel.enabled = YES;
    }
    
    eKMKeyboardState kbState = [[KMManager sharedInstance] stateForKeyboardWithID:keyboardID];
    [(KeyboardNameTableViewCell *)cell setKeyboardState:kbState selected:NO defaultAccessoryType:cell.accessoryType];
}

/*
 // Override to support conditional editing of the table view.
 - (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
 {
 // Return NO if you do not want the specified item to be editable.
 return YES;
 }
 */

/*
 // Override to support editing the table view.
 - (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
 {
 if (editingStyle == UITableViewCellEditingStyleDelete) {
 // Delete the row from the data source
 [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
 }
 else if (editingStyle == UITableViewCellEditingStyleInsert) {
 // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
 }
 }
 */

/*
 // Override to support rearranging the table view.
 - (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath
 {
 }
 */

/*
 // Override to support conditional rearranging of the table view.
 - (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath
 {
 // Return NO if you do not want the item to be re-orderable.
 return YES;
 }
 */

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView cellForRowAtIndexPath:indexPath].selected = NO;
    NSString *kbID = [[keyboards objectAtIndex:indexPath.section] objectForKey:kKeymanIdKey];
    NSString *kbName = [[keyboards objectAtIndex:indexPath.section] objectForKey:kKeymanNameKey];
    //NSString *isRTL = [[keyboards objectAtIndex:indexPath.section] objectForKey:kKeymanKeyboardRTLKey];
    //isRTL = (isRTL == nil)?@"N":@"Y";
    //NSString *font = [[KMManager sharedInstance] jsFontFromFontDictionary:[[keyboards objectAtIndex:indexPath.section] objectForKey:kKeymanFontKey]];
    //NSString *oskFont = [[KMManager sharedInstance] jsFontFromFontDictionary:[[keyboards objectAtIndex:indexPath.section] objectForKey:kKeymanOskFontKey]];
    
    eKMKeyboardState state = [[KMManager sharedInstance] stateForKeyboardWithID:kbID];
    if (state != kKMKeyboardStateDownloading) {
        if (state == kKMKeyboardStateNeedsDownload)
            self.isUpdate = NO;
        else
            self.isUpdate = YES;
        
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:[NSString stringWithFormat:@"%@: %@", languageName, kbName]
                                                        message:@"Would you like to download this keyboard?"
                                                       delegate:self
                                              cancelButtonTitle:@"Cancel"
                                              otherButtonTitles:@"Download", nil];
        alert.tag = indexPath.section;
        [alert show];
    }
    /*
    if (state == kKMKeyboardStateNeedsDownload) {
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:[NSString stringWithFormat:@"%@: %@", languageName, kbName]
                                                         message:@"Would you like to download this keyboard?"
                                                        delegate:self
                                               cancelButtonTitle:@"Cancel"
                                               otherButtonTitles:@"Download", nil];
        alert.tag = indexPath.section;
        [alert show];
    }
    else if (state == kKMKeyboardStateDownloading || state == kKMKeyboardStateNone) {
        // Do nothing.
    }
    else {
        // Add keyboard.
        [[KMManager sharedInstance] addKeyboardWithID:kbID languageID:languageID keyboardName:kbName languageName:languageName isRTL:[isRTL isEqualToString:@"Y"]?YES:NO isCustom:NO font:font oskFont:oskFont];
        [[KMManager sharedInstance] setKeyboardWithID:kbID languageID:languageID keyboardName:kbName languageName:languageName font:font oskFont:oskFont];
        [self.navigationController popToRootViewControllerAnimated:YES];
    }*/
}

#pragma mark - Alert view delegate

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {   
    // Keyboard download confirmation alert (tag is used for keyboard index).
    if (buttonIndex != alertView.cancelButtonIndex) {
        [[KMManager sharedInstance] downloadKeyboardForLanguageIndex:languageIndex keyboardIndex:alertView.tag isUpdate:self.isUpdate];
    }
}

#pragma mark - Keyman notifications

- (void)keyboardDownloadStarted:(NSNotification *)notification {
    [self.view setUserInteractionEnabled:NO];
    [self.navigationItem setHidesBackButton:YES animated:YES];
    
    CGRect frame = [self.navigationController.toolbar frame];
    CGFloat width = frame.size.width;
    CGFloat height = frame.size.height;
    frame.size.width = width*0.95;
    frame.size.height = height*0.7;
    UILabel *label = [[UILabel alloc] initWithFrame:frame];
    [label setBackgroundColor:[UIColor clearColor]];
    [label setTextColor:[UIColor whiteColor]];
    [label setTextAlignment:NSTextAlignmentCenter];
    [label setCenter:CGPointMake(width*0.5, height*0.5)];
    [label setText:@"Downloading\u2026"];
    [label setAutoresizingMask:UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleTopMargin |  UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight];
    [label setTag:kmToolbarLabelTag];
    
    UIActivityIndicatorView *indicatorView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
    frame = [indicatorView frame];
    [indicatorView setCenter:CGPointMake(width - frame.size.width, height*0.5)];
    [indicatorView setAutoresizingMask:UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleTopMargin |  UIViewAutoresizingFlexibleBottomMargin];
    [indicatorView setTag:kmToolbarActivityIndicatorTag];
    [indicatorView startAnimating];
    
    [[self.navigationController.toolbar viewWithTag:kmToolbarButtonTag] removeFromSuperview];
    [[self.navigationController.toolbar viewWithTag:kmToolbarLabelTag] removeFromSuperview];
    [[self.navigationController.toolbar viewWithTag:kmToolbarActivityIndicatorTag] removeFromSuperview];
    [self.navigationController.toolbar addSubview:label];
    [self.navigationController.toolbar addSubview:indicatorView];
    [self.navigationController setToolbarHidden:NO animated:YES];
}

- (void)keyboardDownloadFailed:(NSNotification *)notification {
    [self.view setUserInteractionEnabled:YES];
    [self.navigationItem setHidesBackButton:NO animated:YES];
}

#pragma mark - Private methods

- (void)loadUserKeyboards {
    NSUserDefaults *userData = [[KMManager sharedInstance] activeUserDefaults];
    NSArray *userKbList = [[NSArray alloc] initWithArray:[userData arrayForKey:kKeymanUserKeyboardsListKey]];
    NSUInteger len = [userKbList count];
    
    if (len == 0) {
        userKeyboards = nil;
        return;
    }
    
    NSString *langID = nil;
    NSString *kbID = nil;
    NSString *dictKey = nil;
    userKeyboards = [[NSMutableDictionary alloc] initWithCapacity:len];
    for (NSDictionary *kb in userKbList) {
        langID = [kb objectForKey:kKeymanLanguageIdKey];
        kbID = [kb objectForKey:kKeymanKeyboardIdKey];
        dictKey = [NSString stringWithFormat:@"%@_%@", langID, kbID];
        [self.userKeyboards setObject:kb forKey:dictKey];
    }
}

- (BOOL)isAdded:(NSString *)langID keyboardID:(NSString *)kbID {
    if (userKeyboards != nil) {
        if([userKeyboards objectForKey:[NSString stringWithFormat:@"%@_%@", langID, kbID]]) {
            return YES;
        }
        else {
            return NO;
        }
    }
    else {
        return NO;
    }
}

@end

File: ios/engine/KMEI/KeymanEngine/Classes/LanguagePicker/LanguageDetailViewController.swift
//
//  LanguageDetailViewController.swift
//  KeymanEngine
//
//  Created by Gabriel Wong on 2017-09-14.
//  Copyright © 2017 SIL International. All rights reserved.
//

import UIKit

private let toolbarButtonTag = 100
private let toolbarLabelTag = 101
private let toolbarActivityIndicatorTag = 102

class LanguageDetailViewController: UITableViewController, UIAlertViewDelegate {
  private let activityViewTag = -2

  private var userKeyboards: [String: InstallableKeyboard] = [:]
  private var isUpdate = false
  private var language: Language
  private var keyboardRepository: KeyboardRepository?
  private var pendingFetch: Bool = false

  private var keyboardDownloadStartedObserver: NotificationObserver?
  private var keyboardDownloadFailedObserver: NotificationObserver?

  convenience init(language: Language) {
    self.init(nil, language: language)
  }
  
  init(_ keyboardRepository: KeyboardRepository?, language: Language) {
    self.language = language
    super.init(nibName: nil, bundle: nil)
    
    self.keyboardRepository = keyboardRepository
    keyboardRepository?.delegate = self
  }

  required init?(coder aDecoder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  override func loadView() {
    super.loadView()
    
    if keyboardRepository != nil {
      if let languageDict = keyboardRepository?.languages {
        self.postLanguageLoad(languageDict: languageDict)
      } else {
        log.info("Fetching repository from API for keyboard download (LanguageDetailViewController)")
        pendingFetch = true
        keyboardRepository?.fetch()
      }
    }
    
    loadUserKeyboards()
  }
  
  func postLanguageLoad(languageDict: [String: Language]) {
    pendingFetch = false
    
    if let language = languageDict[self.language.id] {
      self.language = language
      
      // Case 1 - called immediately during loadView() - UI is built after the load, so we're fine.
      // Case 2 - called by keyboardRepositoryDidFetch() - UI updated by self.tableView.reloadData()
    }
  }

  override func viewDidLoad() {
    super.viewDidLoad()
    keyboardDownloadStartedObserver = NotificationCenter.default.addObserver(
      forName: Notifications.packageDownloadStarted,
      observer: self,
      function: LanguageDetailViewController.keyboardDownloadStarted)
    keyboardDownloadFailedObserver = NotificationCenter.default.addObserver(
      forName: Notifications.packageDownloadFailed,
      observer: self,
      function: LanguageDetailViewController.keyboardDownloadFailed)
    log.info("viewDidLoad: LanguageDetailViewController (registered for keyboardDownloadStarted)")
  }
  
  override func viewDidAppear(_ animated: Bool) {
    super.viewDidAppear(animated)
    navigationController?.setToolbarHidden(true, animated: true)
    // if no rows to show yet, show a loading indicator
    if pendingFetch {
      showActivityView()
      log.info("didAppear: LanguageDetailViewController, but fetch is not yet complete")
    } else {
      log.info("didAppear: LanguageDetailViewController")
    }
  }

  // MARK: - Table view data source UITableViewDataSource

  override func numberOfSections(in tableView: UITableView) -> Int {
    return language.keyboards!.count
  }

  override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return 1
  }

  override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cellIdentifier = "Cell"
    if let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier) {
      return cell
    }

    let cell = KeyboardNameTableViewCell(style: .subtitle, reuseIdentifier: cellIdentifier)
    let selectionColor = UIView()
    selectionColor.backgroundColor = Colors.selectionPrimary
    cell.selectedBackgroundView = selectionColor
    return cell
  }

  override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
    let keyboard = language.keyboards![indexPath.section]
    let cell = cell as! KeyboardNameTableViewCell
    cell.indexPath = indexPath
    cell.textLabel?.text = keyboard.name
    if isAdded(languageID: language.id, keyboardID: keyboard.id) {
      cell.accessoryType = .checkmark
      cell.isUserInteractionEnabled = false
      cell.textLabel?.isEnabled = false
      cell.detailTextLabel?.isEnabled = false
    } else {
      cell.accessoryType = .none
      cell.isUserInteractionEnabled = true
      cell.textLabel?.isEnabled = true
      cell.detailTextLabel?.isEnabled = true
    }

    let kbState = ResourceDownloadManager.shared.stateForKeyboard(withID: keyboard.id)
    cell.setKeyboardState(kbState, selected: false, defaultAccessoryType: cell.accessoryType)
  }

  override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    tableView.cellForRow(at: indexPath)?.isSelected = false
    let keyboardIndex = indexPath.section
    let keyboard = language.keyboards![keyboardIndex]

    let state = ResourceDownloadManager.shared.stateForKeyboard(withID: keyboard.id)
    if state != .downloading {
      if state == .needsDownload {
        isUpdate = false
      } else {
        isUpdate = true
      }
    
        let alertController = UIAlertController(title: "\(language.name): \(keyboard.name)",
                                                message: "Would you like to download this keyboard?",
                                                preferredStyle: UIAlertController.Style.alert)
        alertController.addAction(UIAlertAction(title: "Cancel",
                                                style: UIAlertAction.Style.cancel,
                                                handler: nil))
        alertController.addAction(UIAlertAction(title: "Download",
                                                style: UIAlertAction.Style.default,
                                                handler: {_ in self.downloadHandler(keyboardIndex)} ))
        
        self.present(alertController, animated: true, completion: nil)
    }
  }

  func downloadHandler(_ keyboardIndex: Int) {
    let fullID = FullKeyboardID(keyboardID: language.keyboards![keyboardIndex].id, languageID: language.id)
    let installClosure = ResourceDownloadManager.shared.standardKeyboardInstallCompletionBlock(forFullID: fullID, withModel: true)
    ResourceDownloadManager.shared.downloadKeyboard(withID: fullID.keyboardID,
                                                    languageID: fullID.languageID,
                                                    isUpdate: isUpdate,
                                                    completionBlock: installClosure)
  }
  
  func showActivityView() {
    view.isUserInteractionEnabled = false
    let indicatorView = UIActivityIndicatorView(style: .whiteLarge)
    let activityView = UIView(frame: indicatorView.bounds.insetBy(dx: -10.0, dy: -10.0))
    activityView.backgroundColor = Colors.spinnerBackground
    activityView.layer.cornerRadius = 6.0
    activityView.center = view.center
    activityView.tag = activityViewTag
    activityView.autoresizingMask = [.flexibleLeftMargin, .flexibleRightMargin, .flexibleTopMargin,
                                     .flexibleBottomMargin]

    indicatorView.center = CGPoint(x: activityView.bounds.size.width * 0.5, y: activityView.bounds.size.height * 0.5)
    indicatorView.startAnimating()
    activityView.addSubview(indicatorView)
    view.addSubview(activityView)
  }

  func dismissActivityView() {
    let activityView = view.viewWithTag(activityViewTag)
    activityView?.removeFromSuperview()
    view.isUserInteractionEnabled = true
  }

  private func showConnectionErrorAlert() {
    Alerts.showConnectionErrorAlert(in: self, handler: Alerts.popToNavigationRootHandler(for: self))
  }

  private func keyboardDownloadStarted() {
    log.info("keyboardDownloadStarted: LanguageDetailViewController")
    view.isUserInteractionEnabled = false
    navigationItem.setHidesBackButton(true, animated: true)
  }

  private func keyboardDownloadFailed() {
    log.info("keyboardDownloadFailed: LanguageDetailViewController")
    Alerts.showDownloadErrorAlert(in: self, handler: Alerts.popToNavigationRootHandler(for: self))

    view.isUserInteractionEnabled = true
    navigationItem.setHidesBackButton(false, animated: true)
  }

  private func loadUserKeyboards() {
    guard let userKbList = Storage.active.userDefaults.userKeyboards, !userKbList.isEmpty else {
      userKeyboards = [:]
      return
    }

    userKeyboards = [:]
    for kb in userKbList {
      let dictKey = "\(kb.languageID)_\(kb.id)"
      userKeyboards[dictKey] = kb
    }
  }

  private func isAdded(languageID: String?, keyboardID: String?) -> Bool {
    guard let languageID = languageID, let keyboardID = keyboardID else {
      return false
    }
    return userKeyboards["\(languageID)_\(keyboardID)"] != nil
  }
}

// MARK: - KeyboardRepositoryDelegate
extension LanguageDetailViewController: KeyboardRepositoryDelegate {
  func keyboardRepositoryDidFetch(_ repository: KeyboardRepository) {
    if let languageDict = repository.languages {
      self.postLanguageLoad(languageDict: languageDict)
    }
    self.dismissActivityView()
    self.tableView.reloadData()
    if self.numberOfSections(in: self.tableView) == 0 {
      self.showConnectionErrorAlert()
    }
  }

  func keyboardRepository(_ repository: KeyboardRepository, didFailFetch error: Error) {
    dismissActivityView()
    showConnectionErrorAlert()
  }

  private func languageList(_ languageDict: [String: Language]) -> [Language] {
    return languageDict.values.sorted { a, b -> Bool in
      a.name.localizedCaseInsensitiveCompare(b.name) == .orderedAscending
    }
  }
}
